keps/sig-windows/1981-windows-privileged-container-support/README.md

```diff
@@ -195,27 +198,28 @@ The lack of privileged container support within the Windows container model has
 
 Furthermore, since kube-proxy cannot be run as a privileged daemonset, it must either be run with a proxy or directly on the host as a service. In the case that it is run as a service, the admin kube config must be stored on the Windows node which poses a security concern. This is also true for networking daemons such as Flannel.
 
-
 ### Goals
 
 <!--
 List the specific goals of the KEP. What is it trying to achieve? How will we
 know that this has succeeded?
 -->
-- To provide a method to build, launch, and run a Windows-based container with privileged access to host resources, including the host network service, devices, disks (including hostPath volumes), etc.
-- To enable access to host network resources for <b>privileged</b> containers and pods with host network mode
 
+- To provide a method to build, launch, and run a Windows-based container with privileged access to host resources, including the host network service, devices, disks (including hostPath volumes), etc.
```

Comment:
It's hard for me to tell what behavior is being coupled to the `hostProcess` field and what is being coupled to the `hostNetwork` field

What happens to windows pods in the following scenarios:
  1. `hostNetwork: true`, `hostProcess: null`
  2. `hostNetwork: true`, `hostProcess: false`
  3. `hostNetwork: true`, `hostProcess: true` (is the code responsible for making the containers run in host network mode looking at the pod's `hostNetwork` field or the `hostProcess` field? are we plumbing the right information so we won't have to redo that logic if hostProcess containers gain the ability to *not* run in hostNetwork mode in the future, as is hinted at later in this doc)
  4. `hostNetwork: false`, `hostProcess: true` (I know this is proposed to be disallowed in validation, but if validation *did* permit this in the future, how would 1.21-generation kubelets and CRI react?)

Reply:
Similarly, it's hard for me to tell if setting `hostProcess: true` _implicitly_ turns on a bunch of host-level stuff we would eventually want individual control over (e.g. hostNetwork, hostPath volumes/devices, shared process space, etc), or if it merely signals that it is acceptable for that host-level stuff to be requested / set up.

The comments about the interactions with other things defined in the pod (like external volume mounts) make it really hard to understand how much of the meaning of the pod spec is altered by setting this field to `true`

Reply:
> What happens today if you run a pod on a Windows node with hostNetwork: true or a hostPath volume?

Right now there is different behavior for dockershim and containerd if hostNetwork = true

With dockershim docker gets instructed to add the container to a docker network named 'host'. It is expected that this network is created on the node during node provisioning and the container will fail to start if it is not preset.

With containerd the container starts but does not get joined to a host network at all. The appropriate fields are passed along to the CRI calls but containerd completely ignores these fields

Part of the work to support these privileged containers for Windows will be to update containerd to properly support host networking scenarios.

Reply:
> Similarly, it's hard for me to tell if setting hostProcess: true implicitly turns on a bunch of host-level stuff we would eventually want individual control over (e.g. hostNetwork, hostPath volumes/devices, shared process space, etc), or if it merely signals that it is acceptable for that host-level stuff to be requested / set up.

I tried to explain that in the **Overview** section with:

`Windows privileged containers will be implemented with Job Objects, a break from the previous container model using server silos. Job objects provide the ability to manage a group of processes as a group, and assign resource constraints to the processes in the job. Job objects have no process or file system isolation, enabling the privileged payload to view and edit the host file system with the correct permissions, among other host resources. The init process, and any processes it launches or that are explicitly launched by the user, are all assigned to the job object of that container. When the init process exits or is signaled to exit, all the processes in the job will be signaled to exit, the job handle will be closed and the storage will be unmounted.`

These `HostProcess` containers are actually processes running on the host. We aren't implicitly turning on any features and also we aren't signaling that it is acceptable for the host-level stuff to be set up. This is radically differently from how things work on Linux today but it is the best way we've been able to support this scenario as a whole.

Reply:
> It's hard for me to tell what behavior is being coupled to the `hostProcess` field and what is being coupled to the `hostNetwork` field
> 
> What happens to windows pods in the following scenarios:
> 
> 1. `hostNetwork: true`, `hostProcess: null`
> 2. `hostNetwork: true`, `hostProcess: false`
> 3. `hostNetwork: true`, `hostProcess: true` (is the code responsible for making the containers run in host network mode looking at the pod's `hostNetwork` field or the `hostProcess` field? are we plumbing the right information so we won't have to redo that logic if hostProcess containers gain the ability to _not_ run in hostNetwork mode in the future, as is hinted at later in this doc)
> 4. `hostNetwork: false`, `hostProcess: true` (I know this is proposed to be disallowed in validation, but if validation _did_ permit this in the future, how would 1.21-generation kubelets and CRI react?)

When using CRI, setting `hostNetwork:true` in the podSpec does a few things. The two most notable things are
- Sets up a voume mount for /etc/hosts 
- Sets NamepaceOption.NamespaceMode = NODE (POD is default)
https://github.com/kubernetes/kubernetes/blob/017d4dd832173dc4f1bc8a6e1e4d519023d19d70/pkg/kubelet/kuberuntime/helpers.go#L307-L312

Currently containerd (the only CRI implementation for Windows) completely ignores the NapespaceMode setting for Windows and always attaches pods to the pod network/namespace. 

With this proposal  `hostProcess` should only toggle the new `host_process` CRI fields and `hostNetwork` should perform all current behaviors.

The extra checks are added to the kubelet / API validation to prevent known-unsupported scenarios (mainly `hostNetwork: true` , `hostProcess: null / false`